_posts/2024-08-07-accessing-kafka-with-gateway-api.md

```diff
@@ -0,0 +1,627 @@
+
+### KIND Cluster Setup
+
+First, let's create the KIND cluster using the following `kind-config.yaml` file:
+
+```
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+- kubeadmConfigPatches:
+  - |
+    kind: InitConfiguration
+    nodeRegistration:
+      kubeletExtraArgs:
+        node-labels: "ingress-ready=true"
+  extraPortMappings:
+  - containerPort: 30992
+    hostPort: 9092
+    protocol: TCP
+```
+
+If you inspect the `kind-config.yaml` file, you will notice the port mapping of `hostPort: 9092` being mapped to `containerPort: 30992`. That means that the port 9092 on your own laptop will be forwarded by docker to port 30992 on the Kubernetes Node (which, in KIND, is just a docker container running on your laptop). We will use this fact when installing Envoy Gateway.
+
+```
+kind create cluster --name strimzi-gw-api --config kind-config.yaml
+```
+
+Next, let's set up Envoy Gateway. First, we will use `helm` to install it.
+
+```
+helm upgrade --install envoygateway oci://docker.io/envoyproxy/gateway-helm \
+    --version v1.0.1 \
+    --namespace envoy-gateway-system \
+    --create-namespace
+```
+
+Once the installation process is complete, we need to deploy a `Gateway` with a specific `GatewayClass` that will listen on the correct `NodePort`s. We'll refer to this `Gateway` later when we create `TLSRoute`s to access our Kafka cluster. Let's apply the following file:
+
+```
+apiVersion: gateway.networking.k8s.io/v1
+kind: GatewayClass
+metadata:
+  name: my-gateway-class
+  namespace: envoy-gateway-system
+spec:
+  controllerName: gateway.envoyproxy.io/gatewayclass-controller
+  parametersRef:
+    group: gateway.envoyproxy.io
+    kind: EnvoyProxy
+    name: my-proxy-config
+    namespace: envoy-gateway-system
+---
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: EnvoyProxy
+metadata:
+  name: my-proxy-config
+  namespace: envoy-gateway-system
+spec:
+  provider:
+    type: Kubernetes
+    kubernetes:
+      envoyDeployment:
+        replicas: 1
+      envoyService:
+        type: NodePort
+        patch:
+          value:
+            spec:
+              ports:
+              # Port 9092 on your laptop gets forwarded to NodePort 30992 on the KIND cluster.
+              - name: kafka-port
+                nodePort: 30992
+                port: 9092
+                protocol: TCP
+                targetPort: 9092
+---
+apiVersion: gateway.networking.k8s.io/v1
+kind: Gateway
+metadata:
+  name: my-gateway
+  namespace: default
+spec:
+  gatewayClassName: my-gateway-class
+  listeners:
+  - name: kafka-listener
+    protocol: TLS
+    port: 9092
+    tls:
+      mode: Passthrough
+```
```

Comment:
i wonder if this would be worth splitting into two blocks ...
* First block with the EnvoyProxy resource that is Envoy specific and creates the Envoy Gateway deployment
* Second with the two Gateway APi resources referencing it.

That would make it more clear which part belongs where and for example how to replace it when using different implementation. But feel free to ignore it if you think having it all in one is better.

Reply:
I think that's a good idea, in my opinion the stuff in the `envoy-gateway-system` namespace should be separate from the stuff in the `default`/`strimzi` namespace. The `GatewayClass` defines a type of `Gateway` (like Ingress and IngressClass for example).

The `EnvoyProxy` is an Envoy Gateway-specific CRD that configures how Envoy Gateway reconciles a specific `Gateway`.

I wanted this blog post to focus mostly on Kafka, but I guess since it's about Strimzi + Gateway API after all I should probably add this. Will put that in my next commit. Thanks Jakub!